retrievalmarket/impl/blockunsealing/blockunsealing_test.go

```diff
@@ -0,0 +1,142 @@
+		checkSuccessLoad(t, loaderWithUnsealing, testdata.RootNodeLnk)
+		unsealer.VerifyExpectations(t)
+	})
+
+	t.Run("when intermediate blockstore does not have block", func(t *testing.T) {
+		t.Run("unsealing success on first ref", func(t *testing.T) {
+			bs := setupBlockStore(t)
+			unsealer := testnodes.NewTestRetrievalProviderNode()
+			unsealer.ExpectUnseal(deal1.SectorID, deal1.Offset, deal1.Length, carData)
+			loaderWithUnsealing := blockunsealing.NewLoaderWithUnsealing(ctx, bs, piece, cio, unsealer.UnsealSector)
+			checkSuccessLoad(t, loaderWithUnsealing, testdata.MiddleMapNodeLnk)
+			unsealer.VerifyExpectations(t)
+		})
+
+		t.Run("unsealing success on later ref", func(t *testing.T) {
+			bs := setupBlockStore(t)
+			unsealer := testnodes.NewTestRetrievalProviderNode()
+			unsealer.ExpectFailedUnseal(deal1.SectorID, deal1.Offset, deal1.Length)
+			unsealer.ExpectUnseal(deal2.SectorID, deal2.Offset, deal2.Length, carData)
+			loaderWithUnsealing := blockunsealing.NewLoaderWithUnsealing(ctx, bs, piece, cio, unsealer.UnsealSector)
+			checkSuccessLoad(t, loaderWithUnsealing, testdata.MiddleMapNodeLnk)
+			unsealer.VerifyExpectations(t)
+		})
+
+		t.Run("fails all unsealing", func(t *testing.T) {
+			bs := setupBlockStore(t)
+			unsealer := testnodes.NewTestRetrievalProviderNode()
+			unsealer.ExpectFailedUnseal(deal1.SectorID, deal1.Offset, deal1.Length)
+			unsealer.ExpectFailedUnseal(deal2.SectorID, deal2.Offset, deal2.Length)
+			loaderWithUnsealing := blockunsealing.NewLoaderWithUnsealing(ctx, bs, piece, cio, unsealer.UnsealSector)
+			_, err := loaderWithUnsealing.Load(testdata.MiddleMapNodeLnk, ipld.LinkContext{})
+			require.Error(t, err)
+			unsealer.VerifyExpectations(t)
+		})
+
+		t.Run("car io failure", func(t *testing.T) {
+			bs := setupBlockStore(t)
+			unsealer := testnodes.NewTestRetrievalProviderNode()
+			randBytes := make([]byte, 100)
+			_, err := rand.Read(randBytes)
+			require.NoError(t, err)
+			unsealer.ExpectUnseal(deal1.SectorID, deal1.Offset, deal1.Length, randBytes)
+			loaderWithUnsealing := blockunsealing.NewLoaderWithUnsealing(ctx, bs, piece, cio, unsealer.UnsealSector)
+			_, err = loaderWithUnsealing.Load(testdata.MiddleMapNodeLnk, ipld.LinkContext{})
+			require.Error(t, err)
+			unsealer.VerifyExpectations(t)
+		})
+
+		t.Run("when piece was already unsealed", func(t *testing.T) {
+			bs := setupBlockStore(t)
+			unsealer := testnodes.NewTestRetrievalProviderNode()
+			unsealer.ExpectUnseal(deal1.SectorID, deal1.Offset, deal1.Length, carData)
+			loaderWithUnsealing := blockunsealing.NewLoaderWithUnsealing(ctx, bs, piece, cio, unsealer.UnsealSector)
+			checkSuccessLoad(t, loaderWithUnsealing, testdata.MiddleMapNodeLnk)
+			// clear out block store
+			err := bs.DeleteBlock(testdata.MiddleMapBlock.Cid())
+			require.NoError(t, err)
+
+			// attemp to load again, will not unseal, will fail
+			_, err = loaderWithUnsealing.Load(testdata.MiddleMapNodeLnk, ipld.LinkContext{})
+			require.Error(t, err)
```

Comment:
yea so, this isn't the best test, but the error is actually the same as if it weren't in the blockstore --

if it unsealed a second time, it would work, cause it would put the deleted block back in the block store.

so it's not a specific error -- it just doesn't attempt another unseal -- and then fails reading from the blockstore as a result